### PR TITLE
Fix language semantics and add chained comparisons

### PR DESCRIPTION
**Pull request by shmolyneaux**
This PR addresses multiple correctness issues and adds support for chained comparisons in Shimlang.

## Summary
Implements 15 fixes to language semantics and behavior, along with comprehensive documentation updates. The changes ensure safer arithmetic, better error handling, and more intuitive comparison syntax.

## Key Changes

**Arithmetic Safety:**
- Integer arithmetic (`+`, `-`, `*`, `pow`) now saturates at `i32::MIN`/`i32::MAX` instead of panicking on overflow/underflow
- Division by zero returns `0` (or `0.0` for floats) instead of `inf`/`NaN`
- Modulo by zero returns `0` instead of panicking

**Chained Comparisons:**
- Added support for chained comparison syntax: `a < b < c` is now equivalent to `a < b and b < c`
- Each operand is evaluated at most once, and evaluation short-circuits on the first false comparison
- Applies to all relational operators (`<`, `<=`, `>`, `>=`, `==`, `!=`)
- Implemented via new `Expression::Compare` variant and `Swap` bytecode instruction

**Numeric Literals:**
- Added support for `e`/`E` exponent notation in float literals (`1.5e3`, `2.5E-2`)
- Added support for `_` digit separators in integer and float literals (`1_000_000`, `1_000.000_5`)
- Added support for leading-dot float literals (`.5` equivalent to `0.5`)

**Error Handling & Validation:**
- Added compile-time validation that `break`/`continue` only appear inside loop bodies; now returns formatted error instead of runtime panic
- Added validation that `break`/`continue` cannot cross function boundaries
- Non-ASCII bytes in string literals now produce a parse error instead of crashing later

**Type System & Built-ins:**
- Added `bool()` built-in function to convert any value's truthiness to a boolean
- Dictionary keys now distinguish between integer and float keys (e.g., `1` and `1.0` are different keys)
- Non-finite float values (`inf`, `-inf`, `NaN`) now format without trailing `.0`

**Formatting & Display:**
- Recursive data structures (lists, tuples, structs) now detect cycles and emit `...` instead of infinite recursion
- Improved error messages to avoid exposing internal debug representations

**Documentation:**
- Updated LANGUAGE.md with comprehensive documentation of all new features and behaviors
- Added test scripts demonstrating new functionality and edge cases
- Added error test cases for compile-time validation

## Implementation Details

- Loop control validation uses recursive traversal of the AST, tracking loop context and resetting at function boundaries
- Chained comparisons are parsed into a single `Compare` expression node containing operands and operators
- Compilation of chained comparisons uses stack duplication (`CopyFrom`) and the new `Swap` bytecode to efficiently reuse operands
- Dictionary key equality now uses a new `dict_key_equal` method that distinguishes types while the `==` operator treats `1 == 1.0` as true
- Numeric literal parsing refactored to handle exponents and separators in a single pass

https://claude.ai/code/session_015XBxdNaV6Fw9TSQp2QaJT1

### CODE REVIEW

File: shimlang/LANGUAGE.md [hunk 1]
```diff
@@ -148,6 +176,40 @@ Output:
...
+
+```
+1500.0
+0.025
+0.5
+1000.0005
+```
+
+Non-finite values are written `inf`, `-inf`, and `NaN` (note the absence of a
+trailing `.0`). They arise from operations such as `sqrt` of a negative number:
```
Comment (shmolyneaux) [inline shimlang/LANGUAGE.md]: Remove the "note the absence of a trailing" part

File: shimlang/src/runtime.rs [hunk 2]
```diff
@@ -1760,16 +1798,65 @@ impl ShimValue {
...
     pub fn div(&self, interpreter: &mut Interpreter, other: &Self) -> Result<ShimValue, String> {
+        // Division always produces a float; division by zero is defined to
+        // return 0.0 rather than yielding infinity/NaN or panicking.
         match (self, other) {
-            (ShimValue::Integer(a), ShimValue::Integer(b)) => Ok(ShimValue::Float((*a as f32) / (*b as f32))),
-            (ShimValue::Float(a), ShimValue::Float(b)) => Ok(ShimValue::Float(*a / *b)),
-            (ShimValue::Integer(a), ShimValue::Float(b)) => Ok(ShimValue::Float((*a as f32) / *b)),
-            (ShimValue::Float(a), ShimValue::Integer(b)) => Ok(ShimValue::Float(*a / (*b as f32))),
+            (ShimValue::Integer(a), ShimValue::Integer(b)) => {
+                if *b == 0 { Ok(ShimValue::Float(0.0)) }
```
Comment (shmolyneaux) [inline shimlang/src/runtime.rs]: Use a `(_, ShimValue::Integer(0) | ShimValue::Float(0.0))` sort-of match arm to handle this

Comment (shmolyneaux) [reply shimlang/src/runtime.rs]: (same with modulo)

**Review by shmolyneaux — COMMENTED** [review]
Work on feedback
